testsuite/manualmode/src/test/java/org/jboss/as/test/manualmode/deployment/DeploymentScannerUnitTestCase.java

```diff
@@ -222,6 +225,47 @@ public void testFailedDeploymentWithPersistentDeployment() throws Exception {
                             () -> !exists(PathAddress.pathAddress(DEPLOYMENT, JAR_TWO)));
                 } finally {
                     removeDeploymentScanner();
+                    client.execute(Util.createRemoveOperation(persistentDeploymentAddress));
+                }
+            } finally {
+                StreamUtils.safeClose(client);
+            }
+        } finally {
+            container.stop();
+        }
+    }
+
+/**
+     * https://issues.jboss.org/browse/WFCORE-1890
+     *
+     * When FS deployment is erplaced with a managed deployment with same name it is not marked as undeployed and reboot will fail.
+     */
+    @Test
+    public void testReplaceDeploymentWithPersistentDeployment() throws Exception {
+        container.start();
+        try {
+            client = TestSuiteEnvironment.getModelControllerClient();
+            try {
+                final PathAddress persistentDeploymentAddress = PathAddress.pathAddress(DEPLOYMENT, JAR_ONE);
+                addDeploymentScanner(0);
+                try {
+                    // deploy an file-system deployment
+                    container.stop();
+                    createDeployment(deployDir.resolve(JAR_ONE), "org.jboss.modules");
+                    container.start();
+                    Path deployedMarker = deployDir.resolve(JAR_ONE + ".deployed");
+                    waitFor(String.format("Missing .deployed marker for %s", JAR_ONE),
+                            () -> Files.exists(deployedMarker));
+                    Assert.assertTrue(String.format("%s should be deployed", JAR_ONE), exists(persistentDeploymentAddress));
+                    //Replace deployment
+                    Archive<?> validDeployment = createDeploymentArchive();
+                    replaceWithPersistent(JAR_ONE, validDeployment);
+                    Assert.assertTrue(String.format("%s should be deployed", JAR_ONE), exists(persistentDeploymentAddress));
+                    Assert.assertTrue(String.format("Missing .undeployed marker for %s", JAR_ONE),
+                            Files.exists(deployDir.resolve(JAR_ONE + ".undeployed")));
```

Comment:
We discussed this in chat; I'm just putting a comment here in case others review the PR. The management notification to the scanner is async to the management op that emits it so the test can't assume it's received and handled immediately after the management op completes. There needs to be some sort allowance for a possible delay.